### DISCUSSION TITLE
Vite + Express + HTTPS not working

### DISCUSSION THREAD

**Discussion by gustavopch** [remix-run/remix#8046] [Q&A] [answered]
Did anyone have success using Vite + Express + HTTPS? I'm looking for some code I can use as reference.

My HTTPS configuration that worked without Vite doesn't seem to be compatible with Vite. The browser can't find the dev server and this happens:
![image](https://github.com/remix-run/remix/assets/10941958/3e48b06f-44a2-4ef2-807c-e5f323498de4)

Would be nice to have that in the docs too and maybe even in https://github.com/remix-run/remix/tree/main/templates or https://github.com/remix-run/examples.

**Comment by hi-ogawa** [comment, marked as answer]
I started from the official [`unstable-vite-express` template](https://github.com/remix-run/remix/blob/main/templates/unstable-vite-express/README.md) and experimented with https here https://github.com/hi-ogawa/test-remix-vite-express-https.
I think I made it work both `dev` (with hmr) and `build + start` (though I'm not particularly experience with using https locally, so I might have missed something).
I hope it helps!
If something still doesn't work, I would appreciate if you can share more details about your setup.

**Reply by gustavopch** [reply]
I ended up doing the same as you did: configuring key and cert both in the server file and in the Vite config. It seems to work!

**Reply by hi-ogawa** [reply]
Just FYI, there was a new insight regarding custom server and https setup.
You won't need extra `server.https.key/cert` config on vite side if you instantiate vite middleware in the following way:

```tsx
import https from "node:https";
import { createServer } from "vite";

const server = https.createServer(
  {
    key: fs.readFileSync("localhost-key.pem"),
    cert: fs.readFileSync("localhost.pem"),
  },
  app
);

let vite = await createServer({
  server: {
    middlewareMode: true,
    hmr: {
      server,        // <=========  user `server.hmr.server`
    }
  }
})
```

See details in here https://github.com/remix-run/remix/pull/8095
